test/object-store/results.cpp

```diff
@@ -3640,6 +3640,96 @@ TEST_CASE("results: snapshots") {
     }
 }
 
+TEST_CASE("results: knnsearch") {
+    InMemoryTestFile config;
+    config.cache = false;
+    config.automatic_change_notifications = false;
+
+    auto r = Realm::get_shared_realm(config);
+    r->update_schema({
+        {"object",
+            {{"id", PropertyType::Int},
+             {"embedding", PropertyType::Array | PropertyType::Float}}}
+    });
+
+    auto table = r->read_group().get_table("class_object");
+    ColKey col_id = table->get_column_key("id");
+    ColKey col_lst = table->get_column_key("embedding");
+    
+    r->begin_transaction();
+    {
+        Obj o1 = table->create_object();
+        o1.set(col_id, 1);
+        Lst<float> lst = o1.get_list<float>(col_lst);
+        lst.add(0.003);
+        lst.add(0.004);
+        lst.add(0.005);
+        lst.add(0.100);
+        lst.add(0.010);
+    }
+    
+    {
+        Obj o1 = table->create_object();
+        o1.set(col_id, 2);
+        Lst<float> lst = o1.get_list<float>(col_lst);
+        lst.add(0.001);
+        lst.add(0.004);
+        lst.add(0.005);
+        lst.add(0.100);
+        lst.add(0.010);
+    }
+    
+    {
+        Obj o1 = table->create_object();
+        o1.set(col_id, 3);
+        Lst<float> lst = o1.get_list<float>(col_lst);
+        lst.add(0.001);
+        lst.add(0.004);
+        lst.add(0.005);
+        lst.add(0.100);
+        lst.add(0.010);
+    }
+    
+    {
+        Obj o1 = table->create_object();
+        o1.set(col_id, 4);
+        Lst<float> lst = o1.get_list<float>(col_lst);
+        lst.add(0.004);
+        lst.add(0.005);
+        lst.add(0.010);
+        lst.add(0.025);
+        lst.add(0.100);
+    }
+    
+    {
+        Obj o1 = table->create_object();
+        o1.set(col_id, 5);
+        Lst<float> lst = o1.get_list<float>(col_lst);
+        lst.add(0.003);
+        lst.add(0.007);
+        lst.add(0.008);
+        lst.add(0.020);
+        lst.add(0.100);
+    }
+    r->commit_transaction();
+
+    SECTION("Single knn query") {
+        Results results(r, table->where());
+        auto v = results.knn_search(col_lst, {0.003, 0.005, 0.010, 0.020, 0.100}, 2);
+        REQUIRE(v.size() == 2);
+        REQUIRE(v.get(0).get<Int>(col_id) == 4);
+        REQUIRE(v.get(1).get<Int>(col_id) == 5);
+    }
+    
+    SECTION("Knn combined with regular query") {
+        Results results(r, table->where().less(col_id, 5));
+        auto v2 = results.knn_search(col_lst, {0.003, 0.005, 0.010, 0.020, 0.100}, 2);
+        REQUIRE(v2.size() == 2);
+        REQUIRE(v2.get(0).get<Int>(col_id) == 4);
+        REQUIRE(v2.get(1).get<Int>(col_id) == 1);
+    }
```

Comment:
it would be great to also have a test for combining filter/distinct/sort. Something like `results.distinct(aFieldWithDuplicates).knn_search(...).sort(col_id, descending)`

Reply:
Btw, does it make sense to allow it actually? After knn results are sorted starting from the nearest. Does it make sense to allow sort/filter/distinct afterwards? Also, i'd assume that this query is rather slow compared to other operations, we don't want to encourage doing it in the middle somewhere, no?